pandas/core/generic.py

```diff
@@ -8426,7 +8426,10 @@ def first(self: FrameOrSeries, offset) -> FrameOrSeries:
             return self
 
         offset = to_offset(offset)
-        end_date = end = self.index[0] + offset
+        if offset._day_opt == "end" and offset.is_on_offset(self.index[0]):
```

Comment:
Could we use rollforward?

Reply:
Instead of the if-else? Then no, This would work for offsets like 1M, but not for something like 10 Days. That is also the thing which lead me to putting ``offset._day_opt == "end"`` in. But we could use it like

```
        if offset._day_opt == "end":
            end_date = end = offset.rollforward(self.index[0])
        else:
            end_date = end = self.index[0] + offset
```

if this would be preferrably

Reply:
this is an OK solution, but ideally there should be a way to do this with only public methods.

> but not for something like 10 Days

can you give an example? (possibly merits a test so i dont try to incorrectly simplify this somewhere down the line)  if its only for Tick offsets, special-casing wouldn't be that bad since we already special case them a few lines down

Reply:
Tick works, thanks.

``test_first_subset`` in `` pandas/tests/frame/methods/test_first_and_last.py `` covers the behavior and avoids the simplification (as I learned myself a bit earlier :))